internal/loader/settings_client.go

```diff
@@ -16,21 +16,77 @@ import (
 
 //go:generate mockgen -destination=mocks/mock_settings_client.go -package mocks . SettingsClient
 
-type SelectorSettingsClient struct {
-	selectors []acpv1.Selector
+type SettingsResponse struct {
+	Settings []azappconfig.Setting
+	Etags    map[acpv1.Selector][]*azcore.ETag
+}
+
+type EtagSettingsClient struct {
+	etags map[acpv1.Selector][]*azcore.ETag
 }
 
 type SentinelSettingsClient struct {
 	sentinel        acpv1.Sentinel
 	etag            *azcore.ETag
 	refreshInterval string
 }
+type SelectorSettingsClient struct {
+	selectors []acpv1.Selector
+}
 
 type SettingsClient interface {
-	GetSettings(ctx context.Context, client *azappconfig.Client) ([]azappconfig.Setting, error)
+	GetSettings(ctx context.Context, client *azappconfig.Client) (*SettingsResponse, error)
 }
 
-func (s *SentinelSettingsClient) GetSettings(ctx context.Context, client *azappconfig.Client) ([]azappconfig.Setting, error) {
+func (s *EtagSettingsClient) GetSettings(ctx context.Context, client *azappconfig.Client) (*SettingsResponse, error) {
+	nullString := "\x00"
+	settingsResponse := &SettingsResponse{
+		Etags: make(map[acpv1.Selector][]*azcore.ETag),
+	}
+	for filter, pageEtags := range s.etags {
+		if filter.KeyFilter != nil {
+			if filter.LabelFilter == nil {
+				filter.LabelFilter = &nullString // NUL is escaped to \x00 in golang
+			}
+			selector := azappconfig.SettingSelector{
+				KeyFilter:   filter.KeyFilter,
+				LabelFilter: filter.LabelFilter,
+				Fields:      azappconfig.AllSettingFields(),
+			}
+
+			conditions := make([]azcore.MatchConditions, 0)
+			for _, etag := range pageEtags {
+				conditions = append(conditions, azcore.MatchConditions{IfNoneMatch: etag})
+			}
+
+			pager := client.NewListSettingsPager(selector, &azappconfig.ListSettingsOptions{
+				MatchConditions: conditions,
+			})
+
+			pageCount := 0
+			for pager.More() {
+				pageCount++
+				page, err := pager.NextPage(context.Background())
+				if err != nil {
+					return nil, err
+				}
+				// etag changed, return the non nil settings
+				if page.ETag != nil {
+					return settingsResponse, nil
+				}
+			}
+
+			if pageCount != len(pageEtags) {
+				return settingsResponse, nil
+			}
+		}
+	}
+
+	// no change in the settings, return nil
+	return nil, nil
```

Comment:
`GetSetting()` should never return nil without any error, I would suggest return nil etags and settings rather than nil response.